internal/verify/verify.go

```diff
@@ -17,40 +17,99 @@ import (
 	"github.com/cerbos/cerbos/internal/util"
 )
 
+const suiteNameWithErrors = "Unknown"
+
 type Config struct {
 	TestsDir string
 	Run      string
 }
 
 type Result struct {
-	Results []SuiteResult `json:"results"`
-	Failed  bool          `json:"-"`
+	Suites []SuiteNode `json:"suites"`
+	Failed bool        `json:"-"`
+}
+
+type SuiteNode struct {
+	File       string           `json:"file"`
+	Status     string           `json:"status"`
+	Name       string           `json:"name"`
+	Principals []*PrincipalNode `json:"principals,omitempty"`
+	Skipped    bool             `json:"skipped,omitempty"`
+	Failed     bool             `json:"failed,omitempty"`
+}
+
+func (sn *SuiteNode) Principal(name string) *PrincipalNode {
+	for _, p := range sn.Principals {
+		if p.Name == name {
+			return p
+		}
+	}
+
+	p := &PrincipalNode{
+		Name:      name,
+		Resources: nil,
+	}
+	sn.Principals = append(sn.Principals, p)
+
+	return p
+}
+
+func (sn *SuiteNode) Add(principal, resource, action string) *ActionNode {
+	return sn.Principal(principal).Resource(resource).Action(action)
+}
+
+type PrincipalNode struct {
+	Name      string          `json:"name"`
+	Resources []*ResourceNode `json:"resources"`
 }
 
-type SuiteResult struct {
-	File    string       `json:"file"`
-	Suite   string       `json:"suite"`
-	Tests   []TestResult `json:"tests"`
-	Skipped bool         `json:"skipped,omitempty"`
-	Failed  bool         `json:"failed,omitempty"`
+func (pn *PrincipalNode) Resource(name string) *ResourceNode {
+	for _, r := range pn.Resources {
+		if r.Name == name {
+			return r
+		}
+	}
+
+	r := &ResourceNode{
+		Name:    name,
+		Actions: nil,
+	}
+	pn.Resources = append(pn.Resources, r)
+
+	return r
 }
 
-type TestName struct {
-	TableTestName string `json:"name"`
-	PrincipalKey  string `json:"principal"`
-	ResourceKey   string `json:"resource"`
+type ResourceNode struct {
+	Name    string        `json:"name"`
+	Actions []*ActionNode `json:"actions"`
+}
+
+func (rn *ResourceNode) Action(name string) *ActionNode {
+	for _, a := range rn.Actions {
+		if a.Name == name {
+			return a
+		}
+	}
+
+	a := &ActionNode{
+		Name: name,
+		Data: &DataNode{},
+	}
+	rn.Actions = append(rn.Actions, a)
+
+	return a
 }
 
-func (r TestName) String() string {
-	return fmt.Sprintf("'%s' for resource '%s' by principal '%s'", r.TableTestName, r.ResourceKey, r.PrincipalKey)
+type ActionNode struct {
+	Data *DataNode `json:"data"`
```

Comment:
Let's call this `details`

 ```suggestion
	Details *DetailsNode `json:"details"`
```